loki/transformations/utilities.py

```diff
@@ -32,10 +35,167 @@
     'sanitise_imports', 'replace_selected_kind',
     'single_variable_declaration', 'recursive_expression_map_update',
     'get_integer_variable', 'get_loop_bounds', 'find_driver_loops',
-    'get_local_arrays', 'check_routine_pragmas'
+    'get_local_arrays', 'check_routine_pragmas', 'remove_duplicate_args',
+    'modify_variable_declarations', 'RemoveDuplicateArgs',
 ]
 
 
+class RemoveDuplicateArgs(Transformation):
+    """
+    Transformation to remove duplicate arguments for both caller
+    and callee. 
+
+    .. warning::
+        this won't work properly for multiple calls to the same routine
+        with differing duplicate arguments
+
+    Parameters
+    ----------
+    recurse_to_kernels : bool, optional
+        Remove duplicate arguments only at the driver level or recurse to
+        (nested) kernels (Default: `True`).
+    rename_common : bool, optional
+        Try to rename duplicate arguments by finding a common pattern
+        in those names (Default: `False`).
+    """
+
+    # This trafo only operates on procedures
+    item_filter = (ProcedureItem,)
+
+    def __init__(self, recurse_to_kernels=True, rename_common=False):
+        self.recurse_to_kernels = recurse_to_kernels
+        self.rename_common = rename_common
+
+    def transform_subroutine(self, routine, **kwargs):
+        role = kwargs['role']
+        if role == 'driver' or self.recurse_to_kernels:
+            remove_duplicate_args(routine, rename_common=self.rename_common)
+
+def remove_duplicate_args(routine, rename_common=False):
+    """
+    Utility to remove duplicate arguments for both caller
+    and callee. 
+
+    .. warning::
+        this won't work properly for multiple calls to the same routine
+        with differing duplicate arguments
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine to be transformed.
+    rename_common : bool, optional
+        Try to rename duplicate arguments by finding a common pattern
+        in those names (Default: `False`).
```

Comment:
This relates to the callee side only, right?

It took me a while to figure out what the effect of this is, maybe we can expand the description like this?
```suggestion
        Try to rename dummy arguments in called routines that received the same argument
        on the caller side, by finding a common name pattern in those names (Default: `False`).
```